### ISSUE TITLE
Exception in thread GANGA_Update_Thread_JobRegistry_Monitor

### ISSUE THREAD

**Issue by ast0815** [ganga-devs/ganga#1635]
Hello,

I seem to have a problem when I create lots of jobs at once:

```
Exception in thread GANGA_Update_Thread_JobRegistry_Monitor:
Traceback (most recent call last):
  File "/usr/lib64/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/opt/ppd/scratch/kdf77245/hk_software/studies/TriggerApplication_SN_direction/env/lib/python3.6/site-packages/ganga/GangaCore/Core/MonitoringComponent/Local_Ganga
MC_Service.py", line 584, in run
    log.debug("%s" % str(self.registry_slice))
  File "/opt/ppd/scratch/kdf77245/hk_software/studies/TriggerApplication_SN_direction/env/lib/python3.6/site-packages/ganga/GangaCore/GPIDev/Lib/Registry/RegistrySlice.py
", line 475, in _display
    vals.append(str(obj._index_cache[display_str])[0:width])
  File "/opt/ppd/scratch/kdf77245/hk_software/studies/TriggerApplication_SN_direction/env/lib/python3.6/site-packages/ganga/GangaCore/GPIDev/Base/Objects.py", line 955, i
n _index_cache
    if self._fullyLoadedFromDisk():
  File "/opt/ppd/scratch/kdf77245/hk_software/studies/TriggerApplication_SN_direction/env/lib/python3.6/site-packages/ganga/GangaCore/GPIDev/Base/Objects.py", line 981, i
n _fullyLoadedFromDisk
    return self._getRegistry().has_loaded(self)
  File "/opt/ppd/scratch/kdf77245/hk_software/studies/TriggerApplication_SN_direction/env/lib/python3.6/site-packages/ganga/GangaCore/Core/GangaRepository/Registry.py", l
ine 711, in has_loaded
    return self.repository.isObjectLoaded(obj)
  File "/opt/ppd/scratch/kdf77245/hk_software/studies/TriggerApplication_SN_direction/env/lib/python3.6/site-packages/ganga/GangaCore/Core/GangaRepository/GangaRepository
XML.py", line 1196, in isObjectLoaded
    _id = next(id_ for id_, o in self._fully_loaded.items() if o is obj)
  File "/opt/ppd/scratch/kdf77245/hk_software/studies/TriggerApplication_SN_direction/env/lib/python3.6/site-packages/ganga/GangaCore/Core/GangaRepository/GangaRepository
XML.py", line 1196, in <genexpr>
    _id = next(id_ for id_, o in self._fully_loaded.items() if o is obj)
RuntimeError: dictionary changed size during iteration
```

After this error, the jobs also do not get updated from "submitted" to "running" despite them actually running on the Local backend. The error arises *before* submitting the jobs though.

**Comment by ast0815** [comment]
Here is the code that creates the jobs by the way:

```Python
#!/bin/env ganga

# Explicitly tell ganga to include local files
import sys
if '.' not in sys.path:
    sys.path.append('.')

from more_itertools import chunked

from scripts import sim_ids

# How to divide the jobs
input_ids = sim_ids.simulation_ids
files_per_job = 1

# Create jobs
my_jobs = []
for chunk in chunked(input_ids, files_per_job):
    # Job: run WCSim simulations
    j = Job(name="wcsim")
    j.application = Executable()
    j.application.exe = File("scripts/env_wrapper.sh")
    j.application.args = ['./run_wcsim.py'] + chunk
    j.inputfiles = [LocalFile("scripts/*.py"), LocalFile("WCSim.mac"), LocalFile("macros/*.mac")]
    my_jobs.append(j)
```

I run it from within ganga with `%run -i job_script.py`.

**Comment by egede** [comment]
Hi Lukas,

Does the same thing happen if you use `runfile('job_script.py')` instead of the iPython `%run`? My understanding of `% run` in iPython is that it creates a new process. This means that you have two processes running is parallel which we know can be problematic and create race conditions (as you see here). One of the Google Summer of Coding project we have this summer will hopefully address the issue. `runfile` on the other hand executes inside the same process.

As a secondary comment, it seems to me that you try to replicate the work of a `splitter` in Ganga. Have you taken a look at `ArgSplitter` in Ganga? That will create a single job that has subjobs instead. Should be much more manageable.

```
arglist = []
for chunk in chunked(input_ids, files_per_job):
    arglist.append(['./run_wcsim.py'] + chunk)

# Create a job with subjobs
j = Job(name="wcsim")
j.application = Executable()
j.application.exe = File("scripts/env_wrapper.sh")
j.splitter = ArgSplitter(args=arglist)
```

**Comment by ast0815** [comment]
Hello Ulrik,

I will try the `runfile` command.

As for the ArgSplitter, I am later submitting the jobs to different backends (currently `Condor` and `Local`), depending on where (and when) I have free resources. Doing it with a list of jobs like this seemed like the easiest way to achieve that.

**Comment by ast0815** [comment]
Calling the script with `runfile` yields the same error message. Also I checked and at least from what I can see in `htop` it does not seem like `%run` spawns a new process.

**Comment by egede** [comment]
Regarding `%run` you are correct. I got confused by the documentation that says "It thus
sees its environment as if it were being run as a stand-alone program".

**Comment by egede** [comment]
OK, Ι have a failing super simple test case now. See #1638. 

Regarding the issue with the status of jobs not updated, I am not sure if it is related (which is why I created a new issue for the exception thrown from the monitoring).

**Comment by ast0815** [comment]
Well, if I submit the jobs after this error message, they never get shown as "running". Do you see a different behaviour?

**Comment by egede** [comment]
Haven't tried. I only managed to trigger the problem twice. Then ran many more times and failed. I will try just to fix the problem.

**Comment by egede** [comment]
Thinking about it. When the thread `GANGA_Update_Thread_JobRegistry_Monitor` raises an exception, the thread will stop and no new jobs will be fed into the monitoring. Quitting Ganga and restarting should fix it while you wait for the fix for the underlying problem.

**Comment by ast0815** [comment]
Yeah, that is what I've been doing.

**Comment by egede** [comment]
Will be solved by #1638
